spec.md

```diff
@@ -179,6 +212,32 @@ physically present in the final circuit. Refer to the description of the
 instance statement for details on how to instantiate a module
 ([@sec:instances]).
 
+### Public Modules
+
+A public module is a module that is marked with the `public`{.firrtl} keyword.
+This means that this is a module that is usable external to this circuit.  A
+public module thereby has a defined ABI---knowing the FIRRTL representation is
+sufficient to know its compiled representation.
+
+A public module has a number of restrictions:
+
+1. A public module may not be instantiated in the current circuit.
```

Comment:
because of this point 1, i feel like instead of calling this "public", 'root' would be a better word. That captures more the meaning. We may decide that public is applicable to things that do not meet this requirement 1, later.

Reply:
We can just remove this requirement. I think Will, you, and others have all convinced me that this isn't useful to enforce. I'm planning to back this out and then add language around the types of optimizations that are allowed here. Specifically, a public module that is instantiated must be compiled as if it were not instantiated. However, a module that instantiates a public module is allowed to optimize through a public module.

E.g., it is legal for a parent to inline a public module that it instantiates. However, that public module must still be produced following the Verilog ABI.

Initial circuit:
```
circuit Foo:
  public module Foo:
    inst bar of Bar

  public module Bar:
    wire w: UInt<1>
```

Final legal circuit:

```
circuit Foo:
  public module Foo:
    wire bar_w: UInt<1>
  public module Bar:
    wire w: UInt<1>
```

E.g., if a parent instantiates a public module and drives a port to a constant that is fed out through an output port, then the parent module can constant prop through the instance. However, the constant cannot be moved into the instance.

Initial circuit:

```
circuit Foo:
  public module Foo:
    output b: UInt<1>
    inst bar of bar
    connect bar.a, UInt<1>(1)
    connect b, bar.b
  public module Bar:
    input a: UInt<1>
    output b: UInt<1>
    connect b, a
```

Optimized circuit:
```
  public module Foo:
    output b: UInt<1>
    connect b, UInt<1>(1)
  public module Bar:
    input a: UInt<1>
    output b: UInt<1>
    connect b, a
```

Illegally optimized circuit:
```
  public module Foo:
    output b: UInt<1>
    connect b, UInt<1>(1)
  public module Bar:
```

Reply:
Ah, if you don't enforce it then all @dtzSiFive questions I said didn't matter start mattering again ;) But i think that the rule that a public module has to exist in a verilog file of that name squashes many of the questions.

Reply:
> Specifically, a public module that is instantiated must be compiled as if it were not instantiated. However, a module that instantiates a public module is allowed to optimize through a public module.

Putting this in my own words:
- Optimizations on a module A that instantiates a public module B MAY treat B as if it were equivalent to an extmodule, but optimizations to A MAY also fully consider the contents of B and perform any modifications to A desired (including not instantiating B at all).

- When a module A instantiates a public module B, optimizations on B must ignore any information known about its context in A

Reply:
i am a little bit worried about the first bullet, sicne it might imply that you are actually essentially duplicating the public module into a "not public" copy and using that, which might not get the desired effect.

Reply:
in the future we could add a thing at the instantiation site that is `inst foo of Foo as public` which means "use the public version of this and I mean it" to block the optimizations the other way.

Reply:
Spec currently contains this language:

> Modules have the property that instances can always be *inlined* into the parent
module without affecting the semantics of the circuit.

Inlining I think usually means replacing a thing with its definition, not necessarily deleting the original.  Often the latter is kinda assumed since often after inlining it everywhere the definition is now dead.  But as-is I'd read this to mean it's fine to indeed copy any module's definition out (?) to replace an instance.  With or without public -- but with public it'll never be dead so just want to call this out a bit.

You're absolutely right that the first bullet, by allowing optimizing A using B's definition is essentially equivalent to allowing (at least in some senses) inlining B's definition entirely. :+1:

Maybe we need some way to express/manage that the "identity" (and count) of certain things matter, as well as perhaps their position in some way.  For example, probably don't want to be cloning registers around? But dunno.

Pinning this down would help a lot I think, and has been run into on other related fronts.

Anyway good points, personally I think this should be allowed (copying B into A) and if it's not we should figure out why precisely and pin that down.